### PR TITLE
Blog post for CP ecosystem

### PR DESCRIPTION
**Pull request by dourouc05**

### CODE REVIEW

File: _posts/2021-09-18-cp-0-6.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
+layout: post
+title: "ConstraintProgrammingExtensions 0.6 is released"
+date: 2021-09-18
+categories: [releases]
+permalink: /blog/cpe-0.6-release/
+---
+
+JuMP and MathOptInterface are mostly oriented towards mathematical optimisation.
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:9]:
```suggestion
JuMP and MathOptInterface are oriented towards mathematical optimization. 
```

File: _posts/2021-09-18-cp-0-6.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+constraint programming (CP). Bridging the gap between MOI and constraint programming
+is the goal of the 
+[ConstraintProgrammingExtensions.jl](https://github.com/dourouc05/ConstraintProgrammingExtensions.jl)
+package. 
+
+Discussions for support of constraint programming in the JuMP ecosystem started
+[in 2019](https://github.com/jump-dev/JuMP.jl/issues/2014). The original scope
+was quite limited, as it only planned to propose some modeling capabilities to 
+JuMP. However, quite soon, it morphed into a larger proposal of a full 
+[constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227)
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:20]:
```suggestion
[constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227),
```

File: _posts/2021-09-18-cp-0-6.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+accessible from the same API as JuMP. It also meant giving access to CP solvers. 
+
+Mathematical optimisation models are only based on equality and inequality constraints,
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:23]: This isn't quite correct in MOI. How about 

```
Traditional mathematical optimization models focus on (in)equality constraints,
```

File: _posts/2021-09-18-cp-0-6.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+JuMP. However, quite soon, it morphed into a larger proposal of a full 
+[constraint programming environment](https://github.com/jump-dev/JuMP.jl/issues/2227)
+accessible from the same API as JuMP. It also meant giving access to CP solvers. 
+
+Mathematical optimisation models are only based on equality and inequality constraints, 
+typically linear or convex, with an objective function. Constraint programming 
+allows more variety in the kind of constraints that are supported, with a 
+focus on discrete structure: the best-known constraint is probably `all_different`, 
+which models the fact that several integer-valued variables should take different
+values (no two variables are allowed to have the same value). Extremely efficient
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:28]:
```suggestion
values (no two variables are allowed to have the same value). Efficient
```

File: _posts/2021-09-18-cp-0-6.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  
+- Other pure-Julia CP solvers have been released, but they do not yet use the same API:
+  [JuliaConstraints](https://github.com/JuliaConstraints) and 
+  [SeaPerl](https://github.com/corail-research/SeaPearl.jl). 
+  
+This is not the end of the journey. Although MOI and JuMP are nearing their 1.0 
+release, the CP ecosystem has still a long way to go. In the near term, more
+modelling features should be added as extensions to JuMP and more solvers should
+be made accessible. In the longer term, CP modelling will also benefit from 
+the new generation of nonlinear support.
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:55]: You should add you name to this post. Either at the end, or at the top? Something like:
```
# ConstraintProgrammingExtensions 0.6 is released

_Author: Thibaut Cuvelier (@dourouc05)_
```

**Review by odow — COMMENTED** [review]
Nice! A few comments/suggestions.

You might also consider linking to the Youtube videos from JuMP-dev?

**Comment by dourouc05** [comment]
Better this way

**Comment by odow** [comment]
How about now? I changed a couple of sections around

File: _posts/2021-09-18-cp-0-6.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+categories: [releases]
+permalink: /blog/constraint-programming-update/
+---
+
+_Author: Thibaut Cuvelier (@dourouc05)_
+
+JuMP and MathOptInterface are oriented towards traditional mathematical
+optimization, encompassing problem classes such mixed-integer linear programs
+and conic optimization. However, the MathOptInterface API is amenable to other
+kinds of formalism, including constraint programming.
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:14]:
```suggestion
kinds of formalism, including [constraint programming](https://en.wikipedia.org/wiki/Constraint_programming).
```

File: _posts/2021-09-18-cp-0-6.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  Highlights of the ConstraintProgrammingExtensions.jl package include:
+  * set definitions for [dozens of constraint types](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/sets/)
+  * a number of [bridges](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/bridges_sets/),
+    including many bridges to convert constraint programming models into
+    standard mixed-integer programming models
+  * Support for the FlatZinc file format, a standard in constraint programming
+    to exchange models (similar to the LP and MPS formats for linear programming)
+
+- Interfacing several solvers through ConstraintProgrammingExtensions.jl. These
+  include
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:53]:
```suggestion
  include:
```

File: _posts/2021-09-18-cp-0-6.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  * The pure-Julia [ConstraintSolver.jl](https://github.com/Wikunia/ConstraintSolver.jl/)
+  * The industry standard [IBM CPLEX CP Optimizer](https://github.com/dourouc05/CPLEXCP.jl/)
+  * The open-source [Chuffed](https://github.com/dourouc05/Chuffed.jl/)
+  Other pure-Julia constraint solvers have been released, but they do not yet
+  use ConstraintProgrammingExtensions API:
+  * [JuliaConstraints](https://github.com/JuliaConstraints)
+  * [SeaPerl](https://github.com/corail-research/SeaPearl.jl)
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:60]:
```suggestion
  * [SeaPearl](https://github.com/corail-research/SeaPearl.jl)
```

File: _posts/2021-09-18-cp-0-6.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+- A session on constraint programming talks at JuMP-dev 2021
+  * [The design of MiniZinc](https://www.youtube.com/watch?v=tHi0Jd7vgVA)
+  * [ConstraintProgrammingExtensions.jl](https://www.youtube.com/watch?v=B0lO6HdlFAw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=104)
+  * [ConstraintSolver.jl](https://www.youtube.com/watch?v=FXqQc8bi_vw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=101)
+  * [JuliaConstraints](https://www.youtube.com/watch?v=G4siuvNMj0c&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=4)
+
+This is not the end of the journey. Although MathOptInterface and JuMP are nearing
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:68]:
```suggestion
This is the start of our journey towards constraint programming in JuMP. Although MathOptInterface and JuMP are nearing
```

File: _posts/2021-09-18-cp-0-6.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+In the last few months, the constraint programming ecosystem in JuMP has made
+tremendous progress. Highlights include:
+
+- The v0.6 release of [ConstraintProgrammingExtensions.jl](https://github.com/dourouc05/ConstraintProgrammingExtensions.jl).
+  Highlights of the ConstraintProgrammingExtensions.jl package include:
+  * set definitions for [dozens of constraint types](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/sets/)
+  * a number of [bridges](http://tcuvelier.be/ConstraintProgrammingExtensions.jl/dev/reference/bridges_sets/),
+    including many bridges to convert constraint programming models into
+    standard mixed-integer programming models
+  * Support for the FlatZinc file format, a standard in constraint programming
```
Comment (odow) [inline _posts/2021-09-18-cp-0-6.md:49]:
```suggestion
  * Support for the [FlatZinc file format](https://www.minizinc.org/doc-2.4.3/en/fzn-spec.html), a standard in constraint programming
```

**Comment by dourouc05** [comment]
It's much better now, thanks!

**Review by odow — APPROVED** [review]
LGTM

File: _posts/2021-09-18-cp-0-6.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+title: "An update on constraint programming in JuMP"
+date: 2021-09-18
+categories: [releases]
+permalink: /blog/constraint-programming-update/
+---
+
+_Author: Thibaut Cuvelier (@dourouc05)_
+
+JuMP and MathOptInterface are oriented towards traditional mathematical
+optimization, encompassing problem classes such mixed-integer linear programs
```
Comment (blegat) [inline _posts/2021-09-18-cp-0-6.md:12]: "such" -> "such as"

File: _posts/2021-09-18-cp-0-6.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  * [The design of MiniZinc](https://www.youtube.com/watch?v=tHi0Jd7vgVA)
+  * [ConstraintProgrammingExtensions.jl](https://www.youtube.com/watch?v=B0lO6HdlFAw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=104)
+  * [ConstraintSolver.jl](https://www.youtube.com/watch?v=FXqQc8bi_vw&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=101)
+  * [JuliaConstraints](https://www.youtube.com/watch?v=G4siuvNMj0c&list=PLP8iPy9hna6Q343_8sSq4f306VGLW4TLK&index=4)
+
+This is the start of our journey towards constraint programming in JuMP. Although MathOptInterface and JuMP are nearing
+their 1.0 release, the constraint programming ecosystem has still a long way to go.
+In the near term, more modelling features should be added as extensions to JuMP and
+more solvers should be made accessible. In the longer term, constraint
+programming modelling will also benefit from the next generation of nonlinear
```
Comment (blegat) [inline _posts/2021-09-18-cp-0-6.md:73]: Maybe reference https://github.com/jump-dev/MathOptInterface.jl/issues/846?

**Review by blegat — APPROVED** [review]

**Comment by odow** [comment]
Nice @dourouc05 https://jump.dev/blog/constraint-programming-update/

**Comment by dourouc05** [comment]
Great, thanks :)!

**Pull request merged**